### PR TITLE
Fixed small tree verification issue

### PR DESCRIPTION
**Pull request by Jasleen1**
Adresses #144

### CODE REVIEW

File: akd/src/append_only_zks.rs [hunk 1]
```diff
@@ -315,14 +315,14 @@ impl Azks {
...
-            hash: crate::utils::empty_node_hash::<H>(),
+            label: EMPTY_LABEL,
+            hash: crate::utils::empty_node_hash_no_label::<H>(),
         }; ARITY];
         let state = lcp_node.get_state_at_epoch(storage, epoch).await?;
-
         for (i, child) in state.child_states.iter().enumerate() {
             match child {
                 None => {
+                    println!("i = {}, empty", i);
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:325]: Please stick to `debug!` or `warn!` etc traces. They do proper formatting, and can redirect to other outputs (e.g. file writing) if needed. If this was just for debugging it should get removed before being landed. Same goes for the other println! statements below

File: akd/src/append_only_zks.rs [hunk 2]
```diff
@@ -332,6 +332,7 @@ impl Azks {
                         self.get_latest_epoch(),
                     )
                     .await?;
+                    println!("Label of child {} is {:?}", i, unwrapped_child.label);
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:335]: here

File: akd/src/append_only_zks.rs [hunk 3]
```diff
@@ -341,6 +342,7 @@ impl Azks {
                 }
             }
         }
+        println!("Lcp label = {:?}", longest_prefix);
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:345]: here

File: akd/src/client.rs [hunk 4]
```diff
@@ -61,15 +61,28 @@ pub fn verify_nonmembership<H: Hasher>(
...
+    println!("Lcp real = {:?}", lcp_real);
     for i in 0..ARITY {
         let child_hash = H::merge(&[
             proof.longest_prefix_children[i].hash,
             hash_label::<H>(proof.longest_prefix_children[i].label),
         ]);
         lcp_hash = H::merge(&[lcp_hash, child_hash]);
-        lcp_real = lcp_real.get_longest_common_prefix(proof.longest_prefix_children[i].label);
+        let curr_label = proof.longest_prefix_children[i].label;
+        // if curr_label != EMPTY_LABEL {
```
Comment (slawlor) [inline akd/src/client.rs:77]: Is this check necessary or was just a debugging statement?

File: akd/src/directory.rs [hunk 5]
```diff
@@ -666,7 +666,7 @@ mod tests {
         auditor::audit_verify,
         // BUG: Key history verification is failing on small-sized trees. It's being addressed in Issue #144
```
Comment (slawlor) [inline akd/src/directory.rs:667]: If we're really solving this issue, we should remove notes to Issue #144 in the code comments

File: akd/src/directory.rs [hunk 6]
```diff
@@ -681,10 +681,12 @@ mod tests {
     async fn test_empty_tree_root_hash() -> Result<(), AkdError> {
         let db = AsyncInMemoryDatabase::new();
         let vrf = HardCodedAkdVRF {};
-        let akd = Directory::<_, _>::new::<Blake3>(&db, &vrf, false).await?;
+        let akd = Directory::<_, _>::new::<Blake3_256<BaseElement>>(&db, &vrf, false).await?;
```
Comment (slawlor) [inline akd/src/directory.rs:684]: nit: Why? `Blake3` is much more concise

File: akd/src/directory.rs [hunk 7]
```diff
@@ -681,10 +681,12 @@ mod tests {
...
 
         let current_azks = akd.retrieve_current_azks().await?;
-        let hash = akd.get_root_hash::<Blake3>(&current_azks).await?;
+        let hash = akd
+            .get_root_hash::<Blake3_256<BaseElement>>(&current_azks)
```
Comment (slawlor) [inline akd/src/directory.rs:688]: nit: Here too, `Blake3` is just cleaner as a type alias

File: akd/src/directory.rs [hunk 8]
```diff
@@ -740,136 +741,137 @@ mod tests {
     }
 
     // BUG: Key history verification is failing on small-sized trees. It's being addressed in Issue #144
```
Comment (slawlor) [inline akd/src/directory.rs:743]: If solved, we should remove this

File: akd/src/node_state.rs [hunk 9]
```diff
@@ -747,11 +747,11 @@ mod tests {
 
     #[test]
     pub fn serialize_deserialize() {
-        use winter_crypto::hashers::Blake3_256;
+        use winter_crypto::hashers::Sha3_256;
```
Comment (slawlor) [inline akd/src/node_state.rs:750]: Why are we changing the hashing scheme here?

File: akd/src/node_state.rs [hunk 10]
```diff
@@ -747,11 +747,11 @@ mod tests {
...
         use winter_crypto::Hasher;
         use winter_math::fields::f128::BaseElement;
 
-        type Blake3 = Blake3_256<BaseElement>;
+        type Blake3 = Sha3_256<BaseElement>;
```
Comment (slawlor) [inline akd/src/node_state.rs:754]: This isn't correct, if we are going to use SHA3 256, we should make the alias something else or revert back to blake3

**Review by slawlor — CHANGES_REQUESTED** [review]
In `directory.rs` lines 1104-1111 should also be uncommented and tested as they were failing due to this issue

Additionally a MySQL test is failing, so we'll need to see what is going on. It's another LCP hash <> hash error, so it seems related

File: akd/src/client.rs [hunk 11]
```diff
@@ -61,15 +61,28 @@ pub fn verify_nonmembership<H: Hasher>(
     proof: &NonMembershipProof<H>,
 ) -> Result<bool, AkdError> {
     let mut verified = true;
-    let mut lcp_hash = H::hash(&[]);
+    let mut lcp_hash = H::hash(&EMPTY_VALUE);
     let mut lcp_real = proof.longest_prefix_children[0].label;
+    // if lcp_real == EMPTY_LABEL {
```
Comment (slawlor) [inline akd/src/client.rs:66]: Also are the necessary or just debugging statements no longer needed?

File: akd/src/directory.rs (hunk 6 above)
Comment (Jasleen1) [reply akd/src/directory.rs:684]: This particular test has a hard-coded hash to check against. I had in mind that a future PR could replace the concise `Blake3` with `TestHash` (for e.g.), which could depend on a feature or be a simple change of variable somewhere to try different hash functions. This test will fail, if that is not taken care of by actually fixing the hash fn to `Blake3_256<BaseElement>`.

File: akd/src/directory.rs (hunk 7 above)
Comment (Jasleen1) [reply akd/src/directory.rs:688]: Same response as above for this test.

**Pull request merged**